File: src/utils/geocodeService.js
const fs = require('fs');
const path = require('path');

/**
 * Merges multiple geocode objects into a single object with unique keys.
 * @param {Array<Object>} geocodeObjects - Array of geocode objects from different regions.
 * @returns {Object} - A merged object with unique country codes as keys.
 */
function mergeGeocodeObjects(geocodeObjects) {
  // Start with an empty result object
  const mergedGeocode = {};
  
  // Iterate through each object and merge its properties
  geocodeObjects.forEach(obj => {
    if (obj && typeof obj === 'object' && !Array.isArray(obj)) {
      Object.assign(mergedGeocode, obj);
    }
  });
  
  return mergedGeocode;
}

/**
 * Reads the geocode file for a single region.
 * @param {string} region - Name of the region.
 * @param {string} basePath - Path of the geocodes folder.
 * @returns {Promise<Object>} - Promise that resolves to the geocode object for the region.
 */
function readGeocodeForRegion(region, basePath) {
  return new Promise((resolve) => {
    const filePath = path.join(basePath, `${region}-codes.json`);
    fs.readFile(filePath, 'utf8', (err, data) => {
      if (err) {
        console.error(`Error reading geocode file for ${region}: ${err.message}`);
        return resolve({});
      }
      try {
        const geocodes = JSON.parse(data);
        resolve(geocodes);
      } catch (parseErr) {
        console.error(`Error parsing geocode file for ${region}: ${parseErr.message}`);
        resolve({});
      }
    });
  });
}

/**
 * Returns merged geocodes for the given regions, with unique country codes as keys.
 * @param {Array<string>} regions - Array of region names.
 * @param {string} basePath - Path to the geocodes folder.
 * @returns {Promise<Object>} - Promise that resolves to the merged geocodes object.
 */
function getMergedGeocodes(regions, basePath) {
  const readPromises = regions.map(region => readGeocodeForRegion(region, basePath));
  return Promise.all(readPromises)
    .then(geocodeObjects => mergeGeocodeObjects(geocodeObjects));
}

module.exports = { getMergedGeocodes };

File: src/utils/geocode_service.py
#!/usr/bin/env python3
import json
from pathlib import Path
from lite_logging.lite_logging import log

def merge_geocode_objects(geocode_objects: list[dict]) -> dict:
    """
    Merges multiple geocode objects into a single object with unique keys.
    
    Args:
        geocode_objects (list): List of geocode objects from different regions.
    
    Returns:
        dict: A merged object with unique country codes as keys.
    """
    # Start with an empty result object
    merged_geocode = {}
    
    # Iterate through each object and merge its properties
    for obj in geocode_objects:
        if obj and isinstance(obj, dict) and not isinstance(obj, list):
            merged_geocode.update(obj)
    
    return merged_geocode

def read_geocode_for_region(region: str, base_path: str) -> dict:
    """
    Reads the geocode file for a single region.
    
    Args:
        region (str): Name of the region.
        base_path (str): Path of the geocodes folder.
    
    Returns:
        dict: The geocode object for the region.
    """
    file_path = Path(base_path) / f"{region}-codes.json"
    
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            return json.load(f)
    except FileNotFoundError:
        log(f"Error reading geocode file for {region}: File not found", level="ERROR")
        return {}
    except json.JSONDecodeError as e:
        log(f"Error parsing geocode file for {region}: {str(e)}", level="ERROR")
        return {}
    except Exception as e:
        log(f"Error reading geocode file for {region}: {str(e)}", level="ERROR")
        return {}

def get_merged_geocodes(regions: list[str], base_path: str) -> dict:
    """
    Returns merged geocodes for the given regions, with unique country codes as keys.
    
    Args:
        regions (list): List of region names.
        base_path (str): Path to the geocodes folder.
    
    Returns:
        dict: The merged geocodes object.
    """
    geocode_objects = [read_geocode_for_region(region, base_path) for region in regions]
    return merge_geocode_objects(geocode_objects)